UI/CombatReport/CombatLogWnd.cpp

```diff
@@ -66,52 +66,145 @@ class CombatLogWnd::Impl {
+
+    /// Selects ships and populated planets in \a objects and stores them in collections
+    /// for each empire.
+    std::map<int, EmpireForces> ForcesByOwner(const std::set<int>& owners,
+                                              const std::set<int>& objects) {
+        std::map<int, EmpireForces> forces_by_owner;
+
+        for (int object_id : objects) {
+            // gets destroyed objects, so this function can be used for initial combat forces even
+            // if some are destroyed during the combat
+            auto object = Objects().Object(object_id);
+            if (ShouldObjectBeShownInEmpireForces(object)) {
+                forces_by_owner[object->Owner()].push_back(object);
             }
         }
-        return objects_per_owner;
+
+        return forces_by_owner;
     }
 
-    // converts input empire ids and numbers of objects to text with names and
-    // links to empires and numbers-as-text of how many objects they own
-    std::string CountsToText(const std::map<int, int>& count_per_empire,
-                             const std::string& delimiter = ", ")
-    {
+    std::string EmpireIdToText(int empire_id) {
+        if (const Empire* empire = GetEmpire(empire_id))
+            return GG::RgbaTag(empire->Color()) + "<" + VarText::EMPIRE_ID_TAG + " " +
+                   std::to_string(empire->EmpireID()) + ">" + empire->Name() + "</" +
+                   VarText::EMPIRE_ID_TAG + ">" + "</rgba>";
+        else
+            return GG::RgbaTag(ClientUI::DefaultLinkColor()) + UserString("NEUTRAL") + "</rgba>";
+    }
+
+    /// converts to "Empire_name: n" text
+    std::string CountToText(int empire_id, int forces_count) {
         std::stringstream ss;
-        for (auto it = count_per_empire.begin(); it != count_per_empire.end(); ) {
-            std::string owner_string;
-            if (const Empire* owner = GetEmpire(it->first))
-                owner_string = GG::RgbaTag(owner->Color()) + "<" + VarText::EMPIRE_ID_TAG + " "
-                    + std::to_string(owner->EmpireID()) + ">" + owner->Name()
-                    + "</" + VarText::EMPIRE_ID_TAG + ">" + "</rgba>";
-            else
-                owner_string = GG::RgbaTag(ClientUI::DefaultLinkColor()) + UserString("NEUTRAL")  + "</rgba>";
-            ss << owner_string << ": " << it->second;
-            ++it;
-            if (it != count_per_empire.end())
-                ss << delimiter;
+        ss << EmpireIdToText(empire_id) << ": " << forces_count;
+        return ss.str();
+    }
+
+    /// Used to sort empire foces for display
+    class ForcesSorter {
+    public:
+        ForcesSorter(int viewing_empire_id_)
+            : viewing_empire_id(viewing_empire_id_) {
+            }
+                    
+        bool operator()(const std::shared_ptr<UniverseObject>& lhs,
+                        const std::shared_ptr<UniverseObject>& rhs) {
+            if (TypeComparison(lhs->ObjectType(), rhs))
```

Comment:
The `TypeComparison` function seems unnecessary and hard to read... Why not just:
````
if (lhs->ObjectType() < rhs->ObjectType())
    return true;
if (rhs->ObjectType() < lhs->ObjectType())
    return false;
````
or
````
if (lhs->ObjectType() != rhs->ObjectType())
    return lhs->ObjectType() < rhs->ObjectType();
````

Reply:
This was the initial implementation, but the partitioning code later required the same criteria to be used in `ForcesToText`.

```c++
                std::upper_bound(first_of_type, forces_sorted.end(), (*first_of_type)->ObjectType(),
                                 ForcesSorter::TypeComparison);
```

For now it's just comparing types by numerical value, but should another order be desired (planets first, for example), then one change will apply to both sorting and partitioning.

Reply:
The one param as object and the other as the ObjectType still seems odd, though...

Reply:
I agree. As I will be making some changes either way, I'll create a class or function that performs sorting and partitioning in one place.